ddtrace/_trace/span.py

```diff
@@ -619,20 +621,36 @@ def set_link(
         if attributes is None:
             attributes = dict()
 
-        if span_id in self._links:
+        self._set_link_object(
+            SpanLink(
+                trace_id=trace_id,
+                span_id=span_id,
+                tracestate=tracestate,
+                flags=flags,
+                attributes=attributes,
+            )
+        )
+
+    def _set_link_object(self, link: SpanLink) -> None:
+        # We will be changing this behavior to allow certain kinds of span
+        # links to coexist in the _links list even if they have the same
+        # span_id. For now, we are basically reimplementing the old
+        # dictionary-like behavior.
+
+        try:
+            existing_link_idx_with_same_span_id = [link.span_id for link in self._links].index(link.span_id)
```

Comment:
Is this likely to be on a hot path? Creating a list and then searching within it might be quite expensive if done in e.g. a tight loop

Reply:
seems pretty unlikely. afaik span links are used pretty sparingly, with only a small number per span. though i wonder if there's a way to get larger-scale statistics about that.

my context here is span pointers and the initial use cases we have for them. there the most we'll have is 25 pointers on a bulk-item-dynamodb api call. and that's pretty rare (and unlikely to coexist with other span links, so we won't even hit this list stuff per the current implementation of https://github.com/DataDog/dd-trace-py/pull/10621 )